src/main/scala/bench/Benchmark.scala

```diff
@@ -0,0 +1,239 @@
+      measure method "memory-footprint" in {
+        using(generator).curve("memory-footprint") in identity
+      }
+    }
+
+  benchmark("scala.List", scalaLists)
+  benchmark("List", lists)
+  benchmark("LazyList", lazyLists)
+  benchmark("ArrayBuffer", arrayBuffers)
+  benchmark("ListBuffer", listBuffers)
+
+}
+
+trait Cons extends Bench.ForkedTime with Generators {
+
+  def benchmark[A](name: String)(block: ((Int, Object)) => Any): Unit =
+    performance of name in {
+      measure method "cons" in {
+        using(sizeAndObjs).curve("cons") in block
+      }
+    }
+
+  benchmark("scala.List") { case (size, obj) =>
+    var xs = scala.List.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs = obj :: xs
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("List") { case (size, obj) =>
+    var xs = List.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs = obj :: xs
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("LazyList") { case (size, obj) =>
+    var xs = LazyList.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs = obj #:: xs
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("ArrayBuffer") { case (size, obj) =>
+    val xs = mutable.ArrayBuffer.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs += obj
+      i += 1
+    }
+    xs
+  }
+
+  benchmark("ListBuffer") { case (size, obj) =>
+    val xs = mutable.ListBuffer.empty[Object]
+    var i = 0
+    while (i < size) {
+      xs += obj
+      i += 1
+    }
+    xs
+  }
+
+}
+
+trait Uncons extends Bench.ForkedTime with Generators {
+
+  def benchmark[A](name: String, generator: Gen[A])(block: A => Any): Unit =
+    performance of name in {
+      measure method "tail" in {
+        using(generator).curve("uncons") in block
+      }
+    }
+
+  benchmark("scala.List", scalaLists)(_.tail)
+  benchmark("List", lists)(_.tail)
+  benchmark("LazyList", lazyLists)(_.tail)
+  benchmark("ArrayBuffer", arrayBuffers)(_.tail)
+  benchmark("ListBuffer", listBuffers)(_.tail)
+
+}
+
+trait Concat extends Bench.ForkedTime with Generators {
+
+  def benchmark[A](name: String, generator: Gen[A])(block: A => Any): Unit =
+    performance of name in {
+      measure method "concat" in {
+        using(generator).curve("concat") in block
+      }
+    }
+
+  benchmark("scala.List", scalaLists)(xs => xs ++ xs)
+//  concat("List", lists)(xs => xs ++ xs) // Commented because our implementation is not stack-safe
+  benchmark("LazyList", lazyLists)(xs => xs ++ xs)
+  benchmark("ArrayBuffer", arrayBuffers)(xs => xs ++ xs)
+  benchmark("ListBuffer", listBuffers)(xs => xs ++ xs)
+
+}
+
+trait Foreach extends Bench.ForkedTime with Generators {
+
+  def benchmark[A](name: String, generator: Gen[A])(block: A => Any): Unit =
+    performance of name in {
+      measure method "foreach" in {
+        using(generator).curve("foreach") in block
+      }
+    }
+
+  benchmark("scala.List", scalaLists) { xs =>
+    var x: Object = null
+    xs.foreach(x = _)
```

Comment:
Doesn't ScalaMeter do a BlackHole-like thing to consume the results of a benchmark?  It doesn't matter if the JIT compiler inlines the foreach because it will do that in production code, too, if it can.  But it _does_ matter if it can throw away the answer and thus throw away all the work.